picamera2/picamera2.py

```diff
@@ -1035,10 +1034,9 @@ def _execute_or_dispatch(self, function, wait, signal_function):
                 raise RuntimeError("Failure to wait for previous operation to finish!")
             self._dispatch_functions([function], signal_function)
             if self.completed_requests:
-                done, result = function()
-                if done:
-                    self.functions = []
-                    self._future.set_result(result)
+                result = function()
+                self.functions = []
+                self._future.set_result(result)
```

Comment:
Note for @davidplowman. If you want the behaviors here to match the other execution point, it might make sense to wrap the exceptional case into  a future here too? Not totally clear to me.